lib/WWW/YouTube/Download.pm

```diff
@@ -442,6 +442,216 @@ sub user_id {
+
+    print "ok \n" if $self->{verbose};
+
+    return $res->decoded_content;
+}
+
+sub find_playlist_videos {
+	my $self = shift;
+	my $html = shift;
+
+	## find JSON blob
+	my $line;
+	for( split("\n", $html) ){
+		if($_ =~ /window\["ytInitialData"\]/){
+			$line = $_;
+			last;
+		}
+	}
+
+	die "Error: could not find YT JSON data!" unless $line;
+
+	my $start = index($line, '{');
+	my $end = rindex($line, '}');
+	my $json = substr($line, $start, $end - $start + 1);
+
+	my $ref = JSON()->new->utf8(0)->decode($json);
+
+	## drill into YT's data structure
+	die "Error: YT JSON data not structed as expected!" unless
+	$ref->{contents} &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer} &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs} && ref($ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}) eq 'ARRAY' &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0] &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer} &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content} &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer} &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer}->{contents} && ref($ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer}->{contents}) eq 'ARRAY' &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer}->{contents}->[0] &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer}->{contents}->[0]->{itemSectionRenderer} &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer}->{contents}->[0]->{itemSectionRenderer}->{contents} && ref($ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer}->{contents}->[0]->{itemSectionRenderer}->{contents}) eq 'ARRAY' &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer}->{contents}->[0]->{itemSectionRenderer}->{contents}->[0] &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer}->{contents}->[0]->{itemSectionRenderer}->{contents}->[0]->{playlistVideoListRenderer} &&
+	$ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer}->{contents}->[0]->{itemSectionRenderer}->{contents}->[0]->{playlistVideoListRenderer}->{contents} && ref($ref->{contents}->{twoColumnBrowseResultsRenderer}->{tabs}->[0]->{tabRenderer}->{content}->{sectionListRenderer}->{contents}->[0]->{itemSectionRenderer}->{contents}->[0]->{playlistVideoListRenderer}->{contents}) eq 'ARRAY'
```

Comment:
Couldn't we limit this to the final check? Maybe temporarily disable autovivification for this block of code?

Reply:
Mh, correct me if I'm wrong... But I do this step-by-step probing not because I want to prevent autovivification but in order to prevent an error where (potentially) on accessing a deep-level element an upper level elem/key is not defined at all or not of expected type. Is there an easier way I'm not aware of?

Reply:
I see what you're saying. https://metacpan.org/pod/JSON::Path might be pretty easy to implement and should require much less code. I think something like https://metacpan.org/pod/JSON::Validator might do the trick, but may be more work.

Reply:
Still, I would opt against it. JSON::Path looks reasonably sane in it's own right but I feel it's a slew of dependencies for little gain over my code behemoth...